### PR TITLE
[5.4] Optimize UrlGenerator->isValidUrl()

### PR DESCRIPTION
**Pull request by vlakoff**
`UrlGenerator->to()` and `UrlGenerator->asset()` are quite slow. Investigating about this, I found out `UrlGenerator->validUrl()` was a bottleneck, taking by itself about half of the total execution time.

Then, I found out this method could be made significantly faster:

```php
$path = 'foobar';
$nb = 10000;

$t1 = microtime(true);
for ($i = $nb; $i--; ) {
    if (Str::startsWith($path, ['#', '//', 'mailto:', 'tel:', 'http://', 'https://'])) {}
}
$t2 = microtime(true);
for ($i = $nb; $i--; ) {
    if (preg_match('~^(#|//|https?://|mailto:|tel:)~', $path)) {}
}
$t3 = microtime(true);

echo $t2 - $t1, "\n", $t3 - $t2;
```

Before: 673 ms
After: 76 ms

### CODE REVIEW

File: src/Illuminate/Routing/UrlGenerator.php [hunk 1]
```diff
@@ -444,7 +444,7 @@ public function format($root, $path)
      */
     public function isValidUrl($path)
     {
-        if (! Str::startsWith($path, ['#', '//', 'mailto:', 'tel:', 'http://', 'https://'])) {
+        if (! preg_match('~^(#|//|https?://|mailto:|tel:)~', $path)) {
```
Comment (Modelizer) [inline src/Illuminate/Routing/UrlGenerator.php]: it should include `http` as well?

Comment (ntzm) [inline src/Illuminate/Routing/UrlGenerator.php]: It does, `https?://` matches `https://` and `http://`

**Comment by vlakoff** [comment]
I investigated why the http[s] protocols had been added (https://github.com/laravel/framework/commit/ec1b1ae72c241c0d0fa576f18e2615b533379423) while the `filter_var` below supposedly supports them already (and does naive checks thus is quite permissive).

So, I guess it's to allow internationalized domain names, which are not supported by `filter_var()`:
```php
filter_var('http://스타벅스코리아.com', FILTER_VALIDATE_URL) // false
```

**Comment by fernandobandeira** [comment]
It should work with ``` php-intl ``` enabled...

Anyway shouldn't we consider making this check more strict? Currently it seems that ``` http://anything ``` will pass even if theres no ``` .tld ``` there...

**Comment by vlakoff** [comment]
It should be fine as it is. This method is not really for proper validation, but just to determine if the argument is a route or an URL, so that the developer can pass both to e.g. `url()` without caring.

**Pull request merged**